### DISCUSSION TITLE
Contact convergence issues with FunctionNeumannBC

### DISCUSSION THREAD

**Discussion by Flolaffel** [idaholab/moose#20703] [Q&A General] [answered]
Hello,

I'm currently working on a simulation where I want to press a wheel onto the road and then let it roll along. Since I have convergence issues I tried to simplify the problem by only using a segment of the wheel to press on the road so the solve is faster. When using a `FunctionDirichletBC` to press the wheel on the road everything is working fine but as soon as I switch to `FunctionNeumannBC` by system is not converging at all anymore. What could the reason be for this? My input is:

[wheel.txt](https://github.com/idaholab/moose/files/8408642/wheel.txt)

**Comment by recuero** [comment, marked as answer]
Given your description, I wonder if you could be getting rigid body modes in your solution when using Neumann BCs (If problems with convergence are happening when the bodies come into contact). If this happens to be the case, you could use MOOSE control system to switch between BCs once contact is robustly established.

One other thing you could try if the above doesn't work is to lower your contact penalty coefficient in one order of magnitude and assess whether it helps with convergence. 

I have not run your input, these are general ideas.

**Reply by Flolaffel** [reply]
Thanks for your answer. I tried both your suggestions but to no success. When I use the `TimePeriod`-Control to switch from Dirichlet to Neumann the system immediatly loses contact. What am I doing wrong? load_disp is a function establishing contact over time and load_force is just a force that's supposed to hold the contact.

>[Controls]
>  [loading]
>    type = TimePeriod
>    enable_objects = 'BCs::load_disp'
>    disable_objects = 'BCs::load_force'
>    start_time = '0'
>    end_time = '1'
>  []
>[]

Additionally, could you please offer some insight as to why it could be there is no contact detection leading to rigid body movement when using a NeumannBC in the first place? Contact establishes without problems when the simulation is displacement controlled. And I've run tests in MOOSE before where contact detection in force controlled calculations worked too.

**Reply by recuero** [reply]
The way I look at this is to assess whether rigid body modes are likely to occur during the solution process. If you have a square block and take a step using displacement control (as opposed to load control), the intermediate solutions stemming from nonlinear iterations are more likely to have the contact interface to remain in contact. The practical behavior of these effects depends on general boundary conditions applied to the problem and the geometry of it. 

> And I've run tests in MOOSE before where contact detection in force controlled calculations worked too.

Yeah, that makes sense to me. My experience with this behavior is that is problem-dependent.

I think the controls should work in general for this type of problem. Some more suggestions: Use pressure/direct load BCs instead of Neumann BCs, make sure the load you switched to in the `Control`s correspond to the state of deformation you get from the initial displacement conditions, and/or add more boundary conditions to restrict modes of deformation that you are expecting to be restricted by contact interaction.

**Reply by Flolaffel** [reply]
Changing the `NeumannBC` to `Pressure` helped my linear convergence. Sadly my nonlinear convergence is still too bad for the simulation to continue. 

In order to get the value for the Pressure BC I used a `TagVectorAux` AuxKernel and a `NodalSum` Postprocessor to give me the reaction force at the boundary where the DirichletBC is placed and an `AreaPostprocessor` to give me the Area of said boundary. Is that the correct precedure?

Do you have any ideas why there could be convergence issues still?

**Reply by recuero** [reply]
> Sadly my nonlinear convergence is still too bad for the simulation to continue.

One possible alleviating measure is to reduce the penalty coefficient for contact, which would work if that's the root cause.

> Is that the correct procedure?

If you are not using scaling, I think that procedure will give you an approximate value that should be sufficient to make the BC switch. You could of course double check those values by imposing some Dirichlet BCs, instead of contact, on the proper side of the model/mesh.

>  Do you have any ideas why there could be convergence issues still?

If you are still using 1e4 as penalty factor, that may be one possible reason. I forgot to mention the use of the `normalize_penalty` option in the contact action, which would remove the influence of element area when applying the penalty coefficient. I'd definitely play with that.

**Reply by Flolaffel** [reply]
> If you are not using scaling, I think that procedure will give you an approximate value that should be sufficient to make the BC switch. You could of course double check those values by imposing some Dirichlet BCs, instead of contact, on the proper side of the model/mesh.

Ok thanks.

> If you are still using 1e4 as penalty factor, that may be one possible reason. I forgot to mention the use of the normalize_penalty option in the contact action, which would remove the influence of element area when applying the penalty coefficient. I'd definitely play with that.

Reducing the penalty factor doesn't help unluckily. Same goes for the `normalize_penalty` option.

**Reply by GiudGiud** [reply]
@Flolaffel any update on this?
If this is scaling issues, you should scale your problem way down (to like 1000 dof) and look at the conditioning of your matrix using petsc's svd monitor

**Reply by Flolaffel** [reply]
@GiudGiud There is!

I found out that the problem didn't lie in the setup of the physics but rather the mesh. I had to change the number of elements in tangential direction to an even number so there is a set of nodes at the lowest point of the wheel. Only then contact is detected I assume. Furthermore from the displacement controlled simulations there was a 0.001 mm gap between the wheel and the road so there is no initial penetration. That had to be removed, propably for the same reason. 

I added locking correction for accurate results and a predictor for better convergence. The final, converging input file looks like this:

[wheel_new.txt](https://github.com/idaholab/moose/files/8514183/wheel_new.txt)
